### PR TITLE
[ARCTIC-1733] Implement splitable and ordered processes

### PR DESCRIPTION
**Pull request by XBaith**
## Why are the changes needed?

implement #1733 

## Brief change log

1. Split large processes into smaller ones allows you to complete the table optimization and query enhancements to take effect as soon as possible.
2. Try to avoid retries caused by exceptions during the optimization process, such retries may waste resources
3. Order task and process 
    - file size asc: release optimizer resources earier
    - sequence number desc: optimize the latest updated files

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by XBaith** [comment]
cc @wangtaohz @Aireed

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/UnKeyedTableCommit.java [hunk 1]
```diff
@@ -186,14 +189,18 @@ protected void replaceFiles(
       if (CollectionUtils.isNotEmpty(removedDataFiles) ||
           CollectionUtils.isNotEmpty(addedDataFiles)) {
         RewriteFiles dataFileRewrite = transaction.newRewrite();
+        // original snapshot might be expired
+        Snapshot targetSnapshot = table.asUnkeyedTable().snapshot(targetSnapshotId) == null ?
+            table.asUnkeyedTable().currentSnapshot() :
+            table.asUnkeyedTable().snapshot(targetSnapshotId);
```
Comment (wangtaohz) [inline ams/server/src/main/java/com/netease/arctic/server/optimizing/UnKeyedTableCommit.java:195]: The `targetSnapshot` should not be the current snapshot, which cannot achieve the goal of checking the commit conflicts.

Maybe we should open another PR to fix the commit failure issue caused by the expiration of the target snapshot.

Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/optimizing/UnKeyedTableCommit.java:195]: make sense

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/OptimizingQueue.java [hunk 2]
```diff
@@ -140,19 +132,16 @@ public void removeOptimizer(String resourceId) {
...
-
   @Override
   public void ping() {
   }
 
   @Override
   public void touch(String authToken) {
     OptimizerInstance optimizer = getAuthenticatedOptimizer(authToken).touch();
-    LOG.debug("Optimizer {} touch time: {}", optimizer.getToken(), optimizer.getTouchTime());
+    if (LOG.isDebugEnabled()) {
```
[the remark this answers is not in the document]
Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/optimizing/OptimizingQueue.java:142]: I will chery pick this commit

**Comment by shidayang** [comment]
@majin1102 cc

**Comment by wangtaohz** [comment]
I have to make sure when a single plan generates many processes, how should these processes be displayed on the Web UI?

**Comment by wangtaohz** [comment]
Another idea is that **generating only one process per plan** can be seen as a special case of **generating multiple processes per plan**. When users do not mind the cost of planning and want to plan every time to obtain the latest process, this approach can be considered.

**Comment by XBaith** [comment]
> I have to make sure when a single plan generates many processes, how should these processes be displayed on the Web UI?

The mechanism for generating multiple processes is through lazy loading and only executes the next optimizing process until the current process completes. So will not display total processes immediately

**Comment by XBaith** [comment]
Yes, getting a process after planning is a case of getting multiple processes during this implementation

**Comment by wangtaohz** [comment]
> > I have to make sure when a single plan generates many processes, how should these processes be displayed on the Web UI?
> 
> The mechanism for generating multiple processes is through lazy loading and only executes the next optimizing process until the current process completes. So will not display total processes immediately

I think it would be more user-friendly to be able to see all the running processes on the Web UI, which better matches the original design concept. cc @majin1102 @zhoujinsong

**Comment by XBaith** [comment]
> I think it would be more user-friendly to be able to see all the running processes on the Web UI, which better matches the original design concept. cc @majin1102 @zhoujinsong

There is only one **running** process for each table being optimized, other processes will be in the pending state, and a new process will not be created until the previous process is completed, so I think this should meet your original design?

**Pull request closed without merging**